python/private/py_wheel.bzl

```diff
@@ -203,6 +203,427 @@ _DESCRIPTION_FILE_EXTENSION_TO_TYPE = {
 }
 _DEFAULT_DESCRIPTION_FILE_TYPE = "text/plain"
 
+def _escape_filename_distribution_name(name):
+    """Escape the distribution name component of a filename.
+
+    See https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode
+    and https://packaging.python.org/en/latest/specifications/name-normalization/.
+
+    Apart from the valid names according to the above, we also accept
+    '{' and '}', which may be used as placeholders for stamping.
+    """
+    escaped = ""
+    for character in name.elems():
+        if character.isalnum() or character in ["{", "}"]:
+            escaped += character.lower()
+        elif character in ["-", "_", "."]:
+            if escaped == "":
+                fail(
+                    "A valid name must start with a letter or number.",
+                    "Name '%s' does not." % name,
+                )
+            elif escaped.endswith("_"):
+                pass
+            else:
+                escaped += "_"
+        else:
+            fail(
+                "A valid name consists only of ASCII letters ",
+                "and numbers, period, underscore and hyphen.",
+                "Name '%s' has bad character '%s'." % (name, character),
+            )
+    if escaped.endswith("_"):
+        fail(
+            "A valid name must end with a letter or number.",
+            "Name '%s' does not." % name,
+        )
+    return escaped
+
+def normalize_pep440(version):
```

Comment:
It's a parser, yes. If you have ideas on how to make it more approachable, I'm all for that.

My preference would have been to use a regular expression instead — in particular because PEP 440 includes a regular expression that we could take more or less as it is. But regular expressions aren't supported in Starlark, so here I am writing a parser instead. (Though whether a regular expression is more approachable than a parser is perhaps debatable, unless you're really into regular expressions.)

My apologies for the complexity, but PEP 440 is hundreds of lines of English, so some of the complexity seems unavoidable. On the other hand, PEP 440 is a fixed document. Once we have a `normalize_pep440()` function that works to our satisfaction, there will be few (if any) reasons to go back and change it. That's good for maintainability, isn't it, functions that have no conceivable reason to change as the product evolves.